dialogue-target/src/main/java/com/palantir/dialogue/Request.java

```diff
@@ -16,37 +16,203 @@
 
 package com.palantir.dialogue;
 
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSortedMap;
 import com.google.common.collect.Multimap;
+import com.palantir.logsafe.Preconditions;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
-import org.immutables.value.Value;
 
 /** Defines the parameters of a single {@link Call} to an {@link Endpoint}. */
-@DialogueImmutablesStyle
-@Value.Immutable
-public interface Request {
-    /** The HTTP headers for this request, encoded as a map of {@code header-name: header-value}. */
-    Map<String, String> headerParams();
+public final class Request {
+
+    private final ImmutableSortedMap<String, String> headerParams;
+    private final ImmutableMultimap<String, String> queryParams;
+    private final ImmutableMap<String, String> pathParams;
+    private final Optional<RequestBody> body;
+
+    private Request(Builder builder) {
+        body = builder.body;
+        headerParams = builder.headerParams.build();
+        queryParams = builder.queryParams.build();
+        pathParams = builder.pathParams.build();
+    }
+
+    /**
+     * The HTTP headers for this request, encoded as a map of {@code header-name: header-value}.
+     * Headers names are compared in a case-insensitive fashion as per
+     * https://tools.ietf.org/html/rfc7540#section-8.1.2.
+     */
+    public Map<String, String> headerParams() {
+        return headerParams;
+    }
 
     /**
      * The URL query parameters headers for this request. These should *not* be URL-encoded and {@link Channel}s are
      * responsible for performing the appropriate encoding if necessary.
      */
-    Multimap<String, String> queryParams();
+    public Multimap<String, String> queryParams() {
+        return queryParams;
+    }
 
     /**
      * The HTTP path parameters for this request, encoded as a map of {@code param-name: param-value}. There is a
      * one-to-one correspondence between variable {@link PathTemplate.Segment path segments} of a {@link PathTemplate}
      * and the request's {@link #pathParams}.
      */
-    Map<String, String> pathParams();
+    public Map<String, String> pathParams() {
+        return pathParams;
+    }
 
     /** The HTTP request body for this request or empty if this request does not contain a body. */
-    Optional<RequestBody> body();
+    public Optional<RequestBody> body() {
+        return body;
+    }
+
+    @Override
+    public String toString() {
+        return "Request{"
+                + "headerParamsSize="
+                + headerParams.size()
```

Comment:
:+1: nice - no leaking tokens here

Reply:
In a later commit I think I updated this to include header names, not values, for a better debugging experience. Should be in #288